core/src/main/scala/org/apache/spark/scheduler/ShuffleMapTask.scala

```diff
@@ -92,6 +92,12 @@ private[spark] class ShuffleMapTask(
       threadMXBean.getCurrentThreadCpuTime - deserializeStartCpuTime
     } else 0L
 
+    // Register the shuffle write metrics reporter to shuffleWriteMetrics.
+    if (dep.shuffleWriteMetricsReporter.isDefined) {
+      context.taskMetrics().shuffleWriteMetrics.registerExternalShuffleWriteReporter(
```

Comment:
This happens per-task, I think `ShuffleWriteMetrics.externalReporters` can be option instead of array.

Reply:
a simpler idea:
1. create a `class GroupedShuffleWriteMetricsReporter(reporters: Seq[ShuffleWriteMetricsReporter]) extends ShuffleWriteMetricsReporter`, which proxy all the metrics updating to the input reporters.
2. create a `GroupedShuffleWriteMetricsReporter` instance here: `new GroupedShuffleWriteMetricsReporter(Seq(dep.shuffleWriteMetricsReporter.get, context.taskMetrics().shuffleWriteMetrics))`, and pass it to `manager.getWriter`

I think we can use the same approach for read metrics as well.

Reply:
Cool! That's a more cleaner implementation on consistency for both read and write metrics reporter, also read metrics can extend `ShuffleReadMetricsReporter` directly. Done in ca6c407